### PR TITLE
tests: use our own TcpSocket instead of socket2

### PR DESCRIPTION
**Pull request by Keruspe**
Alternative for part of #1372

### CODE REVIEW

**Comment by Thomasdezeeuw** [comment]
@Keruspe I prefer this over my on changes in #1372. Can do three things for me:
* Change the test name to `hup_event_on_disconnect`.
* Add a comment to the `set_linger` call in the test that it is done to prevent a graceful shutdown, per @carllerche comment here: https://github.com/tokio-rs/mio/pull/1372#issuecomment-712899060.
* Pull d77ec615b2544dc061902b2b914896e417abeaaa (or make your own commit).

Then this LGTM.

You can ignore the CI, that is tracked in #1332.

**Review by Thomasdezeeuw — APPROVED** [review]
Thanks.

File: tests/tcp_stream.rs [hunk 1]
```diff
@@ -794,17 +795,17 @@ fn hup() {
 
 #[cfg(windows)]
 fn set_linger_zero(socket: &TcpStream) {
-    use socket2::Socket;
+    use mio::net::TcpSocket;
     use std::os::windows::io::{AsRawSocket, FromRawSocket};
 
-    let s = unsafe { Socket::from_raw_socket(socket.as_raw_socket()) };
+    let s = unsafe { TcpSocket::from_raw_socket(socket.as_raw_socket()) };
```
Comment (carllerche) [inline tests/tcp_stream.rs:801]: Our TCP socket closes on drop, we probably need to `mem::forget` it after setting linger.

**Review by carllerche — CHANGES_REQUESTED** [review]
Thanks, minor request.

**Comment by carllerche** [comment]
fwiw, I added the test threads = 1 after seeing some failure on linux as well.

**Comment by Thomasdezeeuw** [comment]
> fwiw, I added the test threads = 1 after seeing some failure on linux as well.

But then we should just fix the failure(s). The number of concurrent test threads shouldn't influence the tests, and if it would I would consider it a bug.

**Comment by carllerche** [comment]
I don't believe it had to do w/ the test being buggy. I believe it had to do w/ tests overlapping with a defunct TCP connection, i.e. what lingering protects against. I did not do a deep investigation into it, but assuming everything was correct before the PR, this would be the most likely explanation.

File: tests/tcp_stream.rs (hunk 1 above)
Comment (Keruspe) [reply tests/tcp_stream.rs:802]: Done

**Comment by Thomasdezeeuw** [comment]
Ci failure seems unrelated, opened #1376, @carllerche you need to approve it to merge.

**Review by carllerche — APPROVED** [review]
Thanks :+1:

**Comment by Thomasdezeeuw** [comment]
Thanks @Keruspe

**Pull request merged**